flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java

```diff
@@ -105,4 +113,54 @@ public String toString() {
           .toString();
     }
   }
+
+  class CustomCatalogLoader implements CatalogLoader {
+
+    private final SerializableConfiguration hadoopConf;
+    private final Map<String, String> properties;
+    private final String name;
+    private final String impl;
+
+    private CustomCatalogLoader(
+        String name,
+        Map<String, String> properties,
+        Configuration conf,
+        String impl) {
+      this.hadoopConf = new SerializableConfiguration(conf);
+      this.properties = new HashMap<>(properties); // use hashmap for serialization
+      this.name = name;
+      this.impl = Preconditions.checkNotNull(impl,
+          "Cannot initialize custom Catalog because impl property is not set");
+    }
+
+    @Override
+    public Catalog loadCatalog() {
+      DynConstructors.Ctor<Catalog> ctor;
+      try {
+        ctor = DynConstructors.builder(Catalog.class)
+            .impl(impl, Map.class, Configuration.class) // take in flink properties and hadoop configs
+            .impl(impl) // fall back to no-arg constructor
```

Comment:
We definitely need an option to create a catalog without passing `Configuration` but still passing config properties. I originally thought that it would make sense to use another constructor, but then I thought about how `name` is passed... and I think the number of possible constructors may get out of hand.

Instead of adding a lot of constructors, I think we should do this:
1. Use a no-arg constructor for all catalogs
2. Add an `initialize` method to catalogs that is called to pass the catalog name and a string map of config (this matches what Spark does)
3. If the catalog implements Hadoop's `Configurable` interface, also call `setConf` to set the Hadoop config.

That way, we avoid having `Configuration` in any of our APIs and minimize the number of constructors that we need to support. What do you think, @jackye1995?

Reply:
Yes I am also thinking about this issue. Another way I am considering is to use a constructor that only takes a string map. Becasue Hadoop configuration implements the `Iterable<Map.Entry<String, String>>` interface, we can merge Spark or Flink properties with Hadoop configurations together and pass into the constructor in a single map. We can use a wrapper class for the merged map to ensure the names of those properties and config keys do not conflict.

Reply:
I don't think it is a good idea to pass a `Configuration` as a flattened map. Then it isn't possible to reconstruct the original `Configuration` without everything being an override. I think we should keep `Configuration` separate and discourage its use where it isn't needed for Hadoop classes used by Iceberg.

Reply:
Actually, ignore the last comment, I overlooked the fact that the HadoopIO still needs the Hadoop configuration object. So the `iceberg-hive-metastore` already provides a way to solve this, that is to dynamically construct a `CatalogLoader` object, and call `loader.load(conf)`. ([link](https://github.com/apache/iceberg/blob/master/mr/src/main/java/org/apache/iceberg/mr/Catalogs.java#L198))

This provides maximum flexibility, and allow each engine to have a different loader interface that takes different arguments in constructor. Flink already has one, and I can add another one for Spark.

Reply:
> I don't think it is a good idea to pass a `Configuration` as a flattened map. Then it isn't possible to reconstruct the original `Configuration` without everything being an override. I think we should keep `Configuration` separate and discourage its use where it isn't needed for Hadoop classes used by Iceberg.

yeah I replied at the same time lol. Please take a look at the latest reply

Reply:
I don't think that we want to use `CatalogLoader`. The use case for that is a bit different: it is for Hive, where `Configuration` is the _correct_ way to pass options. That loader should call whatever dynamic loader function we introduce in this PR. I also think that it shouldn't be located in the iceberg-hive-metastore module. That should be in iceberg-mr.

Reply:
Oh yes my typo, it's in `iceberg-mr`. So if my understanding is correct, we will have something like:

```java
public class MyCatalog implements Catalog, Configurable {
  
  private Configuration conf;

  public MyCatalog() {
  }
  
  // need to add a default in Catalog interface
  @Override
  public void initialize(Map<String, String> options) {
      // initialize logic
  }
  
  @Override
  public void setConf(Configuration conf) {
    this.conf = conf;
  }
}

// in SparkCatalog
String impl = options.get("impl");
DynConstructors.Ctor<Catalog> ctor = DynConstructors.builder(Catalog.class)
    .impl(impl) // fall back to no-arg constructor
    .buildChecked();
Catalog customCatalog = ctor.newInstance();
if (customCatalog instanceof Configurable) {
  ((Configurable) customCatalog).setConf(conf);
}
customCatalog.initialize(options);
```

Does this look good to you?

Reply:
Mainly, yes. I think `initialize` should be `initialize(String name, Map<String, String> options)` so that we can set the name the catalog is configured to use.